services/src/main/java/org/keycloak/services/resources/KeycloakApplication.java

```diff
@@ -346,37 +344,34 @@ public void importAddUser() {
 
                 for (RealmRepresentation realmRep : realms) {
                     for (UserRepresentation userRep : realmRep.getUsers()) {
-                        KeycloakSession session = sessionFactory.create();
-
-                        try {
+                        try (KeycloakSession session = sessionFactory.create()) {
                             session.getTransactionManager().begin();
                             RealmModel realm = session.realms().getRealmByName(realmRep.getRealm());
 
                             if (realm == null) {
                                 ServicesLogger.LOGGER.addUserFailedRealmNotFound(userRep.getUsername(), realmRep.getRealm());
                             }
 
-                            UserProvider users = session.users();
-
-                            if (users.getUserByUsername(realm, userRep.getUsername()) != null) {
-                                ServicesLogger.LOGGER.notCreatingExistingUser(userRep.getUsername());
-                            } else {
-                                UserModel user = users.addUser(realm, userRep.getUsername());
-                                user.setEnabled(userRep.isEnabled());
-                                RepresentationToModel.createCredentials(userRep, session, realm, user, false);
-                                RepresentationToModel.createRoleMappings(userRep, user, realm);
-                                ServicesLogger.LOGGER.addUserSuccess(userRep.getUsername(), realmRep.getRealm());
+                            try {
+                                UserProvider users = session.users();
+
+                                if (users.getUserByUsername(realm, userRep.getUsername()) != null) {
+                                    ServicesLogger.LOGGER.notCreatingExistingUser(userRep.getUsername());
+                                } else {
+                                    UserModel user = users.addUser(realm, userRep.getUsername());
+                                    user.setEnabled(userRep.isEnabled());
+                                    RepresentationToModel.createCredentials(userRep, session, realm, user, false);
+                                    RepresentationToModel.createRoleMappings(userRep, user, realm);
+                                    ServicesLogger.LOGGER.addUserSuccess(userRep.getUsername(), realmRep.getRealm());
+                                }
+                            } catch (Exception e) {
+                                session.getTransactionManager().setRollbackOnly();
+                                throw e;
                             }
-
-                            session.getTransactionManager().commit();
                         } catch (ModelDuplicateException e) {
-                            session.getTransactionManager().rollback();
                             ServicesLogger.LOGGER.addUserFailedUserExists(userRep.getUsername(), realmRep.getRealm());
```

Comment:
This is not going to be triggered since the session/tx is closed later

Reply:
Will be resolved by using standard `KeycloakModelUtils.runJobInTransaction` mechanism.

Reply:
Relying on KeycloakModelUtils (which is beyond messy and an anti pattern IMO) is not great here. I'd honestly rather then have some method on the session factory or something that does this.